modules/consensus/persist.go

```diff
@@ -1,136 +1,92 @@
-	"runtime"
 
 	"github.com/boltdb/bolt"
 
 	"github.com/NebulousLabs/Sia/build"
-	"github.com/NebulousLabs/Sia/encoding"
-	"github.com/NebulousLabs/Sia/modules"
 	"github.com/NebulousLabs/Sia/types"
 )
 
-// initDatabase is run when the database. This has become the true
-// init function for consensus set
-func (cs *ConsensusSet) initSetDB() error {
-	err := cs.db.startConsistencyGuard()
-	if err != nil {
-		return err
-	}
+const (
+	// DatabaseFilename contains the filename of the database that will be used
+	// when managing consensus.
+	DatabaseFilename = "consensus.db"
+)
 
-	// add genesis block
-	err = cs.db.addBlockMap(&cs.blockRoot)
-	if err != nil {
-		return err
-	}
-	err = cs.db.pushPath(cs.blockRoot.Block.ID())
+// loadDB pulls all the blocks that have been saved to disk into memory, using
+// them to fill out the ConsensusSet.
+func (cs *ConsensusSet) loadDB() error {
+	// Open the database - a new bolt database will be created if none exists.
+	db, err := openDB(filepath.Join(cs.persistDir, DatabaseFilename))
 	if err != nil {
 		return err
 	}
+	cs.db = db
 
-	// Set the siafund pool to 0.
+	// Walk through initialization for Sia.
+	var height types.BlockHeight
 	err = cs.db.Update(func(tx *bolt.Tx) error {
-		sfpBucket := tx.Bucket(SiafundPool)
-		return sfpBucket.Put(SiafundPool, encoding.Marshal(types.NewCurrency64(0)))
+		// Check if the database has been initialized.
+		if !dbInitialized(tx) {
+			return cs.initDB(tx)
+		}
+
+		// Check that the genesis block is correct - typically only incorrect
+		// in the event of developer binaries vs. release binaires.
+		genesisID, err := getPath(tx, 0)
+		if build.DEBUG && err != nil {
+			panic(err)
+		}
+		if genesisID != cs.blockRoot.Block.ID() {
+			return errors.New("Blockchain has wrong genesis block, exiting.")
+		}
+
+		height = blockHeight(tx)
+		return nil
 	})
 	if err != nil {
 		return err
 	}
 
-	// Update the siafundoutput diffs map for the genesis block on
-	// disk. This needs to happen between the database being
-	// opened/initilized and the consensus set hash being calculated
-	for _, sfod := range cs.blockRoot.SiafundOutputDiffs {
-		cs.commitSiafundOutputDiff(sfod, modules.DiffApply)
-	}
-
-	// Prevent the miner payout for the genesis block from being spent
-	cs.db.addSiacoinOutputs(cs.blockRoot.Block.MinerPayoutID(0), types.SiacoinOutput{
-		Value:      types.CalculateCoinbase(0),
-		UnlockHash: types.UnlockHash{},
+	// Send all of the existing blocks to subscribers - temporary while
+	// subscribers don't have any persistence for block progress.
+	return cs.db.View(func(tx *bolt.Tx) error {
+		for i := types.BlockHeight(0); i <= height; i++ {
+			// Fetch the processed block at height 'i'.
+			id, err := getPath(tx, i)
+			if build.DEBUG && err != nil {
+				panic(err)
+			}
+			pb, err := getBlockMap(tx, id)
+			if build.DEBUG && err != nil {
+				panic(err)
+			}
+
+			// Send the block to subscribers.
+			cs.updateSubscribers(nil, []*processedBlock{pb})
```

Comment:
I'm concerned that `updateSubscribers` would call other functions like `ChildTarget` that also result in a `View` tx being opened, then you've got a tx being opened in a tx, and that might cause instability. Since they are both `View` txs, I don't think that there would be a problem but I'm not sure.

The stripped down version of the project demonstrates that this isn't the source of the errors I'm seeing.